geoh5py/shared/merging/base.py

```diff
@@ -0,0 +1,276 @@
+
+    @classmethod
+    def extract_data_information(cls, input_entities: list[ObjectBase]) -> tuple:
+        """
+        Extract the information from data of the input entities.
+        :param input_entities: the list of the object to extract the data from.
+        :return: a tuple containing a list of the data information,
+            the unique entity types for each data,
+            the total length of the vertices,
+            and the number of keys to use to define the entity type.
+        """
+        # extract the data from the input entities
+        id_, previous, nb_keys = 0, 0, 1
+        all_data, unique_entity_types = [], []
+        for input_entity in input_entities:
+            entity_unique_entity_types = []
+
+            for data in input_entity.children:
+                if not isinstance(data, Data):
+                    continue
+                if data.name == "Visual Parameters":
+                    continue
+
+                all_data.append(
+                    [
+                        data.values,
+                        (data.entity_type, data.name, id_),
+                        data.association,
+                        previous,
+                        previous + cast(int, input_entity.n_vertices),
+                    ]
+                )
+                entity_unique_entity_types.append((data.entity_type, data.name))
+                unique_entity_types.append((data.entity_type, data.name, id_))
+                id_ += 1
+
+            # define how to define the entity_type
+            nb_keys = cls.compute_nb_keys(entity_unique_entity_types, nb_keys)
+            previous += cast(int, input_entity.n_vertices)
+
+        # create an intermediate array
+        all_data = np.array(all_data, dtype=object)
+
+        # prepare the attributes
+        unique_entity_types = list(
+            dict.fromkeys(
+                [entity_type[:nb_keys] for entity_type in unique_entity_types]  # type: ignore
+            )
+        )
+
+        return all_data, unique_entity_types, previous, nb_keys
+
+    @classmethod
+    def merge_data(cls, input_entities: list[ObjectBase]) -> dict[str, dict]:
+        """
+        Merge the data respecting the entity type, the values, and the association.
+        :param input_entities: the list of objects to merge the data from.
+        :return: a dictionary of data to add to the merged object.
+        """
+
+        # extract the data from the input entities
+        (
+            all_data,
+            unique_entity_types,
+            total_length,
+            nb_keys,
+        ) = cls.extract_data_information(input_entities)
+
+        # prepare the attributes
+        data_array, unique_associations = cls.prepare_attributes(
+            unique_entity_types, all_data, total_length, nb_keys
+        )
+
+        # create the output data
+        return cls.create_add_data_dictionary(
+            data_array, unique_entity_types, unique_associations, nb_keys
```

Comment:
Still very complicated operations for something that just requires creating new arrays in place. Suggest you simply loop through the data, and create a new Data on the outgoing object if the data.name and data.entity_type.name is different. Otherwise, just insert values in order.

Reply:
I can propose code when I get back. I think I have a pretty simple design in mind. Will see if it works...

Reply:
You corrected it in you rPR.
Closing the conversation